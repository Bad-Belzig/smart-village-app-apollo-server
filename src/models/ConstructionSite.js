import mongoose from 'mongoose';

const Location = mongoose.Schema({
  lat: String,
  lon: String
});

const constructionSite = mongoose.Schema({
  category: String,
  cause: String,
  description: String,
  direction: String,
  endDate: String,
  imageUri: String,
  location: Location,
  locationDescription: String,
  restrictions: [String],
  startDate: String,
  title: String
}, {
  timestamps: true
});

export const ConstructionSite = mongoose.model('ConstructionSite', constructionSite);
