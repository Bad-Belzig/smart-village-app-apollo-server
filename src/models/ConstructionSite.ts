import { model, Schema } from 'mongoose';

const Location = new Schema({
  lat: String,
  lon: String,
});

const constructionSite = new Schema(
  {
    category: String,
    cause: String,
    description: String,
    direction: String,
    endDate: String,
    imageUri: String,
    location: Location,
    locationDescription: String,
    restrictions: [String],
    startDate: String,
    title: String,
  },
  {
    timestamps: true,
  },
);

export const ConstructionSite = model('ConstructionSite', constructionSite);
