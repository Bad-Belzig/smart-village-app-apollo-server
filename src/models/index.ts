export * from './OParl';
export * from './Roadwork';
