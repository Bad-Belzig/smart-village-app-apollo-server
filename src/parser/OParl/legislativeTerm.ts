import { isObjectLike } from 'lodash';

import { parseDate } from '../dateParser';

// only the necessary validation for parsing is done here.
// the created database object will be validated after creation before saving
// by using the .validate method that checks against the provided schema
export const parseLegislativeTerm = (json) => {
  if (isObjectLike(json)) {
    const externalId = json.id; // change the id given to us to be the externalId instead
    delete json.id; // remove the id key from the json to ensure we do not preset it for mongo db
    return {
      ...json,
      externalId,
      created: parseDate(json.created), // this value is valid for OParl v1.1
      modified: parseDate(json.modified), // this value is valid for OParl v1.1

      startDate: parseDate(json.startDate),
      endDate: parseDate(json.endDate),
    };
  }
};
