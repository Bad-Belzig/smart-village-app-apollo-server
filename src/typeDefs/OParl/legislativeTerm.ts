import { gql } from 'apollo-server-core';

import { oParlBase } from './oParlBase';

export const legislativeTermGQL = gql`
  extend type Query {
    oParlLegislativeTerms(externalIds: [String!]): [OParlLegislativeTerm!]
  }

  type OParlLegislativeTerm {
    body: OParlBody
    endDate: Date
    name: String
    startDate: Date
    ${oParlBase}
  }
`;
