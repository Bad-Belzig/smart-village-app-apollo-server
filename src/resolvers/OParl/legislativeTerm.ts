import { Body, ILegislativeTerm, LegislativeTerm } from '../../models';
import { findByIds } from '../resolverHelpers';

export const legislativeTermResolvers = {
  Query: {
    oParlLegislativeTerms: (_, args: { externalIds?: string[] }) =>
      args.externalIds
        ? findByIds(args.externalIds, LegislativeTerm)
        : LegislativeTerm.find(),
  },
  OParlLegislativeTerm: {
    body: (parent: ILegislativeTerm) =>
      Body.findOne({ externalId: parent.body }),
  },
};
