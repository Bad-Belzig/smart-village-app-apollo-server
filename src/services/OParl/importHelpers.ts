import { fetch } from 'apollo-env';
import { isArray, isString } from 'lodash';
import { OParlModel } from '../../models';
import { ImportQueue, ImportQueueEntry, ImportType } from './ImportTypes';

// we expect the input to be either a json object already, or a reference to a json object
// if the value is a string we interpret it as a reference and fetch the corresponding json
// otherwise we just return the value, because we assume it already is a json object
export const getJson = async (value: unknown) => {
  if (isString(value)) {
    try {
      const response = await fetch(value);

      if (response.ok) {
        return await response.json();
      }
    } catch (e) {
      console.log(`Error while fetching ${value}:`, e);
      return;
    }
  }

  return { ...(value as Record<string, unknown>) };
};

export const updateOrCreateEntry = async (
  json: Record<string, unknown>,
  // TODO: improve type of parsers. for now ensure that this is at least a function
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  parseJson: (...args: any) => any,
  model: OParlModel,
  addToQueue?: [ImportQueueEntry | ImportQueueEntry[], ImportType][],
  queue?: ImportQueue,
) => {
  const parsedJson = parseJson(json);

  let entry = isString(parsedJson.externalId)
    ? await model.findOne({
        externalId: parsedJson.externalId,
      })
    : undefined;

  if (entry) {
    console.log(
      'entry already exists:',
      isString(entry) ? entry : entry.externalId,
    );
    entry.overwrite(parsedJson);
  } else {
    entry = new model(parsedJson);
  }

  await entry.validate();

  // add list of oparl objects to import queue after the whole entry is validated
  addToQueue?.forEach((list) => {
    if (isArray(list[0])) {
      queue?.add(
        ...list[0].map(
          (entry) => [entry, list[1]] as [ImportQueueEntry, ImportType],
        ),
      );
    }
  });

  return await entry.save();
};
