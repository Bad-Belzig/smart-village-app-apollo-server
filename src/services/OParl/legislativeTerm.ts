import { fetch } from 'apollo-env';

import { LegislativeTerm } from '../../models';
import { parseLegislativeTerm } from '../../parser';

// import flow
// 1. fetch json
// 2. create db object
//   2.1. validation necessary for parsing (done in parser)
//   2.2. parse json
//   2.3. create object
// 3. validate db object
// 4. queue new related objects
// 5. save to db

export const importLegislativeTerm = async (url: string) => {
  try {
    const response = await fetch(url);

    if (response.ok) {
      const json = await response.json();

      // TODO: check for previous existence

      const lt = new LegislativeTerm(parseLegislativeTerm(json));

      await lt.validate();

      // TODO: check for body to import

      return lt.save();
    } else {
      throw new Error(
        `Error while fetching LegislativeTerm from : ${url}! \n response.status: ${response.status}`,
      );
    }
  } catch (e) {
    console.log(e);
  }
};
