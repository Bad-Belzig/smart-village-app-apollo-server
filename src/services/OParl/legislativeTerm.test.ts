import {
  basicImportTest,
  invalidUrlTest,
  setupTestDataBase,
} from '../../helpers';
import { LegislativeTerm } from '../../models';
import { parseLegislativeTerm } from '../../parser';
import { updateOrCreateEntry } from './importHelpers';
import { importLegislativeTerm } from './legislativeTerm';

const helperFuncs = setupTestDataBase('test_legislative_term_import');

const testEID = 'testEID';
const testType = 'testType';
const testBody = 'testBdo';
// TODO: change test url to something within our control
const testUrl =
  'https://www.muenchen-transparent.de/oparl/v1.0/legislativeterm/0';
const testUrl2 =
  'https://sdnetrim.kdvz-frechen.de/rim4883/webservice/oparl/v1.1/body/1/legislativeterm/468';

describe('importing a legislative term', () => {
  beforeAll(helperFuncs.beforeAll);

  it(
    'imports properly for 1.0',
    basicImportTest(importLegislativeTerm, testUrl),
  );

  it(
    'imports properly for 1.1',
    basicImportTest(importLegislativeTerm, testUrl2),
  );

  it(
    'import returns undefined on faulty url',
    invalidUrlTest(importLegislativeTerm),
  );

  it('does create a new entry when updating with a new externalId', async () => {
    await updateOrCreateEntry(
      { id: testEID, type: testType },
      parseLegislativeTerm,
      LegislativeTerm,
    );

    expect(
      await LegislativeTerm.findOne({ externalId: testEID }),
    ).toHaveProperty('type', testType);
  });

  it('does update the entry when updating with an existing externalId', async () => {
    const oldId = (await LegislativeTerm.findOne({ externalId: testEID }))?.id;

    await updateOrCreateEntry(
      { id: testEID, type: testType, body: testBody },
      parseLegislativeTerm,
      LegislativeTerm,
    );

    expect(
      await LegislativeTerm.findOne({ externalId: testEID }),
    ).toHaveProperty('id', oldId);

    expect(
      await LegislativeTerm.findOne({ externalId: testEID }),
    ).toHaveProperty('body', testBody);
  });

  afterAll(helperFuncs.afterAll);
});
