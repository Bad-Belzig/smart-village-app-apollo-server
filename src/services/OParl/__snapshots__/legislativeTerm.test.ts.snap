// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`importing a legislative term imports properly 1`] = `
Object {
  "__v": 0,
  "externalId": "https://www.muenchen-transparent.de/oparl/v1.0/legislativeterm/0",
  "keyword": Array [],
  "name": "Unbekannt",
  "type": "https://schema.oparl.org/1.0/LegislativeTerm",
}
`;
